### PR TITLE
Builder: Check remote when local img platform doesn't match

### LINKED ISSUE #1144: docker run's --platform option is ignored if the image already exists locally
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior

`docker run`'s `--platform` option consistenly uses the platform
passed by the user.

### Actual behavior

Regardeless of the plaform passed by the user, the only one actually
being used is the one downloaded during the first `docker pull`.

### Steps to reproduce the behavior

On an `amd64` host, if I remove, manually pull, and then run a specific
alpine image, I can only run `amd64` (while I expect the second `docker
run` call to return `aarch64` instead):

```
$ docker rmi alpine:3.12.1
Untagged: [...]
$ docker pull alpine:3.12.1
[...]
$ docker run --platform linux/amd64 alpine:3.12.1 uname -m
x86_64
$ docker run --platform linux/arm64/v8 alpine:3.12.1 uname -m
x86_64
```

If I Remove the image, and try running `arm64` directly (letting docker
pull the image itself), I get the expected `aarch64` but then passing
`--platform linux/amd64` has no effect anymore:

```
$ docker rmi alpine:3.12.1
Untagged: [...]
$ docker run --rm --platform linux/arm64/v8 alpine:3.12.1 uname -m
Unable to find image 'alpine:3.12.1' locally
3.12.1: Pulling from library/alpine
[...]
aarch64
$  docker run --rm --platform linux/amd64 alpine:3.12.1 uname -m
aarch64
```

If I handpick the hashes from `docker buildx imagetools inspect alpine:3.12.1`,
I can run either arch just fine:

```
$ docker run --rm docker.io/library/alpine:3.12.1@sha256:d7342993700f8cd7aba8496c2d0e57be0666e80b4c441925fc6f9361fa81d10e uname -m
x86_64
$ docker run --rm docker.io/library/alpine:3.12.1@sha256:fbb820c07896f5c2516167e7146d9938fc82d4b6b1db167defa5b0a7162e4480 uname -m
aarch64
```

**Output of `docker version`:**

```
Docker version 19.03.12, build 48a66213fe
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.3.1-tp-docker)
  app: Docker Application (Docker Inc., v0.8.0)

Server:
 Containers: 249
  Running: 7
  Paused: 0
  Stopped: 242
 Images: 570
 Server Version: 19.03.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.9.0-1-amd64
 Operating System: Debian GNU/Linux bullseye/sid
 OSType: linux
 Architecture: x86_64
 CPUs: 36
 Total Memory: 31.05GiB
 Name: testbench-001
 ID: CULK:2KJY:HBBJ:4EIF:MGUG:IC2N:6PZ6:54AC:PPPQ:6Q3A:MKRX:OZF5
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: sdelafond
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

This is on a physical machine, running Debian sid, and using the
docker-ce packages from the official repository at
https://download.docker.com/linux/debian.

### PR DESCRIPTION
**Pull request by cpuguy83**
fixes https://github.com/docker/for-linux/issues/1144

This fixes an issue where if a build requests an image that already
exists in Docker's image store but does not match the specified build
platform, instead of building using the wrong platform go ahead and
check the remote for the correct platform.

Steps to reproduce issue:

```terminal
$ docker pull --platform=amd64 debian:buster
<output supressed>
$ cat Dockerfile
FROM debian:buster
RUN echo hello
$ docker build --platform=armhf -< Dockerfile
<output supressed>
```

Without this fix, the build invocation will build using the amd64 image
since it is already tagged locally, but this is clearly not what we
want.

With the fix the local image is not used and instead we pull the correct
image.

### CODE REVIEW

**Comment by cpuguy83** [comment]
Any ideas how to write a test for this?

**Review by tonistiigi — APPROVED** [review]

**Comment by thaJeztah** [comment]
Marked as "WIP", because possibly https://github.com/moby/moby/pull/40725 could also address the problem; @cpuguy83 will look further into that

**Comment by cpuguy83** [comment]
Unmarked WIP since the other PR will not have an effect on this.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@cpuguy83 this one we want to backport?

**Comment by tonistiigi** [comment]
I think we shouldn't backport this. It's subtly changing the behavior and although previous was undocumented and new makes more sense it might still be somewhat surprising in a patch release.

**Comment by thaJeztah** [comment]
👍

**Comment by cpuguy83** [comment]
Agreed with @tonistiigi

File: builder/builder-next/adapters/containerimage/pull.go [hunk 1]
```diff
@@ -174,9 +176,16 @@ func (is *imageSource) ResolveImageConfig(ctx context.Context, ref string, opt l
 		// default == prefer local, but in the future could be smarter
 		fallthrough
 	case source.ResolveModePreferLocal:
-		dt, err := is.resolveLocal(ref)
+		img, err := is.resolveLocal(ref)
 		if err == nil {
-			return "", dt, err
+			if img.Architecture != opt.Platform.Architecture || img.Variant != opt.Platform.Variant || img.OS != opt.Platform.OS {
```
Comment (cpuguy83) [inline builder/builder-next/adapters/containerimage/pull.go]: @tonistiigi Should this have checked if the `img.Variant` is even set?

Comment (tonistiigi) [reply builder/builder-next/adapters/containerimage/pull.go]: It could yes

**Pull request merged**